x-pack/plugins/security_solution/common/api/detection_engine/rule_preview/rule_preview.schema.yaml

```diff
@@ -0,0 +1,93 @@
+openapi: 3.0.0
+info:
+  title: Rule preview API endpoint
+  version: '2023-10-31'
+paths:
+  /api/detection_engine/rules/preview:
+    post:
+      x-labels: [serverless, ess]
+      operationId: RulePreview
+      # Code generator changes are required to build a valid request body Zod schema
+      # TypeSpecificCreateProps gets transformed to a discriminated union which
+      # doesn't have `merge` method code generator uses for `allOf`
+      # `preview_rules_route.ts` contains manually written Zod schemas
+      x-codegen-enabled: false
+      summary: Preview rule alerts generated on specified time range
+      tags:
+        - Rule preview API
+      requestBody:
+        description: An object containing tags to add or remove and alert ids the changes will be applied
+        required: true
+        content:
+          application/json:
+            schema:
+              allOf:
+                - type: object
+                  properties:
+                    invocationCount:
+                      type: integer
+                      minimum: 1
+                    timeframeEnd:
+                      type: string
+                      format: date-time
+                  required: [invocationCount, timeframeEnd]
+                - $ref: '../model/rule_schema/rule_schemas.schema.yaml#/components/schemas/SharedCreateProps'
+                - $ref: '../model/rule_schema/rule_schemas.schema.yaml#/components/schemas/TypeSpecificCreateProps'
```

Comment:
Isn't `SharedCreateProps + TypeSpecificCreateProps == RuleCreateProps`?

In the `model/rule_schema/rule_schemas.schema.yaml` we had to solve the same problem, and it looks like the solution was to merge some types first into types like `EqlRule`, `QueryRule`, etc, and then create a `RuleResponse` as a discriminated union.

I'm not sure it's possible to do the same trick here, because we'd need to combine `RuleCreateProps` with some additional request parameters.

As a side note, we wouldn't have this problem if the request schema would be:

```ts
{
  preview: {
    invocationCount: number;
    timeframeEnd: string;
  },
  rule: RuleCreateProps;
}
```

Reply:
> Isn't SharedCreateProps + TypeSpecificCreateProps == RuleCreateProps?

Yes, that's correct. Though using it here won't help fixing code generation. The problem is in a discriminated union. Since a discriminated union has a discriminator which is a some field it's expected that all union parts are objects. Taking this into account making a usual union on top of a discriminated union isn't so a trivial task and Zod doesn't allow it.

I decided to solve this problem by inlining `RuleCreateProps` internals and combining it with rule preview params.

> As a side note, we wouldn't have this problem if the request schema was:

That's true. It could help the schema much more concise.